comms/src/builder/builder_next.rs

```diff
@@ -0,0 +1,446 @@
+    dial_backoff: Option<BoxedBackoff>,
+    shutdown: Shutdown,
+}
+
+impl CommsBuilder<TcpTransport, PlaceholderService<InboundMessage, (), ()>, PlaceholderService<(), (), ()>, ()> {
+    /// Create a new CommsBuilder
+    pub fn new() -> Self {
+        Self {
+            peer_storage: None,
+            listener_address: None,
+            messaging_pipeline: None,
+            node_identity: None,
+            transport: Some(TcpTransport::new()),
+            dial_backoff: Some(Box::new(ExponentialBackoff::default())),
+            executor: None,
+            protocols: None,
+            shutdown: Shutdown::new(),
+        }
+    }
+}
+
+impl<TTransport, TInPipe, TOutPipe, TOutReq> CommsBuilder<TTransport, TInPipe, TOutPipe, TOutReq>
+where
+    TTransport: Transport + Unpin + Send + Sync + Clone + 'static,
+    TTransport::Output: AsyncRead + AsyncWrite + Send + Sync + Unpin + 'static,
+    TOutPipe: Service<TOutReq, Response = ()> + Clone + Send + 'static,
+    TOutPipe::Error: fmt::Debug + Send,
+    TOutPipe::Future: Send + 'static,
+    TInPipe: Service<InboundMessage> + Clone + Send + 'static,
+    TInPipe::Error: fmt::Debug + Send,
+    TInPipe::Future: Send + 'static,
+    TOutReq: Send + 'static,
+{
+    /// Set the runtime handle to use for spawning tasks. If this is not set the handle that is executing
+    /// `CommsBuilder::spawn` will be used, so this will rarely need to be explicitly set.
+    pub fn with_executor(mut self, handle: runtime::Handle) -> Self {
+        self.executor = Some(handle);
+        self
+    }
+
+    /// Set the [NodeIdentity] for this comms instance. This is required.
+    ///
+    /// [OutboundMessagePool]: ../../outbound_message_service/index.html#outbound-message-pool
+    pub fn with_node_identity(mut self, node_identity: Arc<NodeIdentity>) -> Self {
+        self.node_identity = Some(node_identity);
+        self
+    }
+
+    pub fn with_listener_address(mut self, listener_address: Multiaddr) -> Self {
+        self.listener_address = Some(listener_address);
+        self
+    }
+
+    /// Set the peer storage database to use. This is optional.
+    pub fn with_peer_storage(mut self, peer_storage: CommsDatabase) -> Self {
+        self.peer_storage = Some(peer_storage);
+        self
+    }
+
+    /// Set the backoff that [ConnectionManager] uses when dialing peers. This is optional. If omitted the default
+    /// ExponentialBackoff is used. [ConnectionManager]: crate::connection_manager::next::ConnectionManager
+    pub fn with_dial_backoff<T>(mut self, backoff: T) -> Self
+    where T: Backoff + Send + Sync + 'static {
+        self.dial_backoff = Some(Box::new(backoff));
+        self
+    }
+
+    pub fn with_transport<T>(self, transport: T) -> CommsBuilder<T, TInPipe, TOutPipe, TOutReq>
+    where
+        T: Transport + Unpin + Send + Sync + Clone + 'static,
+        T::Output: AsyncRead + AsyncWrite + Send + Sync + Unpin + 'static,
+    {
+        CommsBuilder {
```

Comment:
Can simplify as:
```Rust
CommsBuilder {
    transport: Some(transport),
    ..self
}
```

Reply:
I wish, but unfortunately you can't - `self` has to be the same type to use that syntax but in this case it isn't - the value we're returning is of a different type. There have been proposals in rust to fix this but nothing has made it into the language as yet.